### PR TITLE
Support impersonation with ID tokens

### PR DESCRIPTION
**Pull request by drcapulet**

### CODE REVIEW

File: lib/google/auth/extras.rb [hunk 1]
```diff
@@ -22,38 +23,58 @@ class RefreshNotSupported < StandardError; end
       # A credential that impersonates a service account. For usage with the
       # older style GCP Ruby SDKs from the google-apis-* gems.
       #
+      # The `email_address` of the service account to impersonate may be the exact
```
Comment (malavbhavsar) [inline lib/google/auth/extras.rb:26]: nit - missing `# @param email_address [String]`?

File: lib/google/auth/extras.rb [hunk 2]
```diff
@@ -67,32 +88,47 @@ def impersonated_authorization(email_address:, scope:, base_credentials: nil, de
       #   The list of email address if there are intermediate service accounts that
       #   need to be impersonated using delegation.
       #
-      # @param email_address [String]
```
Comment (malavbhavsar) [inline lib/google/auth/extras.rb:70]: Hmm, was this removal intentional?

Comment (drcapulet) [reply lib/google/auth/extras.rb:70]: Nope, fixed!

File: lib/google/auth/extras/impersonated_credential.rb [hunk 3]
```diff
@@ -44,36 +75,75 @@ def initialize(email_address:, scope:, base_credentials: nil, delegate_email_add
             transform_email_to_name(email)
           end
 
-          @impersonate_lifetime = lifetime
+          if token_type == :id_token
+            @impersonate_include_email = include_email
+          elsif !include_email.nil?
+            raise ArgumentError, 'Can only provide include_email when using target_audience'
```
Comment (malavbhavsar) [inline lib/google/auth/extras/impersonated_credential.rb:81]: FWIU, `token_type == :id_token` means we are using target_audience. Does this and below error need to be rephrased? Or swapped.

```
      # The token type as symbol, either :id_token or :access_token
      def token_type
        target_audience ? :id_token : :access_token
      end
```

Comment (malavbhavsar) [reply lib/google/auth/extras/impersonated_credential.rb:81]: Ah, sorry i think i get this now. this might help?

```
          if token_type == :id_token # using target audience
            @impersonate_include_email = include_email
          elsif !include_email.nil? # not using target audience and include_email is provided.
            raise ArgumentError, 'Can only provide include_email when using target_audience'
          end
```

Comment (malavbhavsar) [reply lib/google/auth/extras/impersonated_credential.rb:81]: And for below

```ruby
          if token_type == :access_token # not using target audience
            @impersonate_lifetime = lifetime
          elsif !lifetime.nil? # using target audience and lifetime is provided
            raise ArgumentError, 'Cannot provide lifetime when using target_audience'
          end
```

**Review by malavbhavsar — APPROVED** [review]
